scripts/linters/pre_commit_linter.py

```diff
@@ -0,0 +1,517 @@
+        """
+        key = (filepath, mode)
+        if key not in self._CACHE_DATA_DICT:
+            with python_utils.open_file(filepath, mode) as f:
+                lines = f.readlines()
+                self._CACHE_DATA_DICT[key] = (''.join(lines), tuple(lines))
+        return self._CACHE_DATA_DICT[key]
+
+
+def _lint_all_files(
+        js_filepaths, ts_filepaths, py_filepaths, html_filepaths,
+        css_filepaths, file_extension_type, verbose_mode_enabled=False):
+    """Run all lint checks.
+
+    Args:
+        js_filepaths: list(str). The list of js filepaths to be linted.
+        ts_filepaths: list(str). The list of ts filepaths to be linted.
+        py_filepaths: list(str). The list of python filepaths to be linted.
+        html_filepaths: list(str). The list of HTML filepaths to be linted.
+        css_filepaths: list(str). The list of CSS filepaths to be linted.
+        verbose_mode_enabled: bool. True if verbose mode is enabled.
+        file_extension_type: list(str). The list of file extensions to be
+            linted.
+
+    Returns:
+        custom_linter: list. Custom lint checks.
+        third_party_linter: list. Third party lint checks.
+    """
+    parent_dir = os.path.abspath(os.path.join(os.getcwd(), os.pardir))
+    config_path_for_css_in_html = os.path.join(
+        parent_dir, 'oppia', '.stylelintrc')
+    config_path_for_oppia_css = os.path.join(
+        parent_dir, 'oppia', 'core', 'templates', 'css', '.stylelintrc')
+    custom_linters = []
+    third_party_linters = []
+
+    js_ts_file_extension = '.js' in file_extension_type or (
+        '.ts' in file_extension_type)
+    py_file_extension = '.py' in file_extension_type
+    html_file_extension = '.html' in file_extension_type
+    css_file_extension = '.css' in file_extension_type
+
+    js_ts_file_paths = js_filepaths + ts_filepaths
+
+    custom_linter = general_purpose_linter.GeneralPurposeLinter(
+        _FILES[file_extension_type], verbose_mode_enabled=verbose_mode_enabled)
+    custom_linters.append(custom_linter)
+
+    if js_ts_file_extension:
+        custom_linter = js_ts_linter.JsTsLintChecksManager(
+            js_filepaths, ts_filepaths,
+            verbose_mode_enabled=verbose_mode_enabled)
+        third_party_linter = js_ts_linter.ThirdPartyJsTsLintChecksManager(
+            js_ts_file_paths, verbose_mode_enabled=verbose_mode_enabled)
+        custom_linters.append(custom_linter)
+        third_party_linters.append(third_party_linter)
+
+    if html_file_extension:
+        custom_linter = html_linter.HTMLLintChecksManager(
+            html_filepaths, verbose_mode_enabled=verbose_mode_enabled)
+        third_party_linter = html_linter.ThirdPartyHTMLLintChecksManager(
+            html_filepaths, verbose_mode_enabled=verbose_mode_enabled)
+        custom_linters.append(custom_linter)
+        third_party_linters.append(third_party_linter)
+
+        third_party_linter = css_linter.ThirdPartyCSSLintChecksManager(
```

Comment:
I'm confused about this. Why are we calling it "third-party files"?

Also, why are we linting third-party files anyway? We should be linting files that we are writing and storing in this codebase, not third-party files from elsewhere.

(This comment applies elsewhere too.)

Reply:
Oh -- I think I see, you are talking about the lint checker being a third-party library!

Don't do that. From the point of view of this function, we just have one lint checker for each type of file. If you want to split into different threads, each filetype-specific linter should handle that. E.g. you could have a function in each filetype-specific linter that returns a list of linters, and then what you do here is call that function, get the linters, and collate all the tasks for the given filetype-specific linter. 

But at this level the function should not be aware of whether CSS files are being linted by one linter or two linters etc.; the details of the internals of the linters should be something it has no knowledge of, and putting all that stuff here leaks context. Does that make sense? The code here should be generic and similar for all the different filetypes.

Reply:
@seanlip Do you want this(https://github.com/oppia/oppia/pull/8656/commits/0523f04c35dbed9f262d829968ef975bc57e9f26#r389225372)

Reply:
@Hudda -- could you please try and provide more context behind your questions? Instead of "Do you want X", perhaps you can explain why you think X is better or why the analysis I provided has errors/oversights. Try and think through some of the design discussions, too.

Reply:
(I also can't tell what your link is pointing to -- for me, it points to a commit that has 3999 additions and 2969 deletions, which is too much to digest. It would be helpful if you could make your question much more specific.)

Reply:
@seanlip I will try to ask specific questions in future and think through the design discussion too.

Reply:
@seanlip As per discussion with @kevinlee12 it is fine to have task handling at the main script because it provides more granular control.

Reply:
Please note, I discussed this with @kevinlee12 and we both agree on the approach. @kevinlee12 has amended his comment already. PTAL and make sure your implementation is in accordance with the current consensus. Thanks!